# Test Phase - 2 of NRSC-Multiple Docs

In [1]:
# Imports
from unstructured.partition.pdf import partition_pdf
from langchain_core.documents import Document
from pathlib import Path
import time
import pandas as pd

e:\The Leo Programmer\Internships\NRSC\Assignments\testing-models-huge-dataset\.venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
# PDF Files Path
pdf_dir = Path("../../data/pdf/nrsc-pdfs/")
all_pdfs = list(pdf_dir.glob("*.pdf"))

# Filter only relevant PDFs
pdf_files = [
    pdf for pdf in all_pdfs 
    if pdf.name.startswith("Chap") 
    or pdf.name.startswith("Bhuvan") 
    or pdf.name.endswith("Bhuvan.pdf") 
    or pdf.name.endswith("bhuvan.pdf")
]

print(f"Total PDFs found: {len(pdf_files)}")
for p in pdf_files:
    print(" -", p.name)

Total PDFs found: 26
 - Bhuvan-NG_2025.pdf
 - Bhuvan-Webinar-English-2026.pdf
 - BhuvanTrainingBrochure_9-11Dec2025_online.pdf
 - BhuvanTrainingBrochure_9-11Mar2026_online.pdf
 - Chap_10_Atmosphere.pdf
 - Chap_11_Cyclones.pdf
 - Chap_12_FloodDisasterManagement_p1.pdf
 - Chap_12_FloodDisasterManagement_p2.pdf
 - Chap_13_AgricultureDroughtMonitoring.pdf
 - Chap_14_Landslide.pdf
 - Chap_15_Earthquake.pdf
 - Chap_16_ForestFireMonitoring.pdf
 - Chap_1_Agriculture.pdf
 - Chap_2_LULC.pdf
 - Chap_3_ForestVegetation_p1.pdf
 - Chap_3_ForestVegetation_p2.pdf
 - Chap_3_ForestVegetation_p3.pdf
 - Chap_4_Soils.pdf
 - Chap_5_Urban.pdf
 - Chap_6_Water_RS.pdf
 - Chap_7_Geosciences.pdf
 - Chap_8_GroundWater.pdf
 - Chap_9_Oceans.pdf
 - InternationalDayofYoga_Short_bhuvan.pdf
 - OCM2%20NDVI%20Filtered%20Product_for_bhuvan.pdf
 - TSC_Bhuvan.pdf


In [3]:
# PDF Ingestion with Unstructured - hi_res strategy
all_elements = []

for pdf_path in pdf_files:
    file_start = time.time()
    
    elements = partition_pdf(
        filename=str(pdf_path),
        extract_images_in_pdf=False,
        strategy="hi_res",
        hi_res_model_name="yolox",
        infer_table_structure=True,
        chunking_strategy="by_title",   # ← unstructured chunks by section title
        max_characters=1500,            # ← max chunk size
        new_after_n_chars=1200,         # ← start new chunk after 1200 chars
        combine_text_under_n_chars=300, # ← combine tiny sections together
        multipage_sections=True,
    )
    
    for el in elements:
        el.metadata.source = pdf_path.name
    
    all_elements.extend(elements)
    
    print(f"---> {pdf_path.name} → {len(elements)} chunks ({time.time()-file_start:.1f}s)")

print(f"\nTotal elements: {len(all_elements)}")

The `max_size` parameter is deprecated and will be removed in v4.26. Please specify in `size['longest_edge'] instead`.


---> Bhuvan-NG_2025.pdf → 12 chunks (152.8s)
---> Bhuvan-Webinar-English-2026.pdf → 5 chunks (21.3s)
---> BhuvanTrainingBrochure_9-11Dec2025_online.pdf → 7 chunks (23.3s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_10_Atmosphere.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> BhuvanTrainingBrochure_9-11Mar2026_online.pdf → 7 chunks (24.6s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_10_Atmosphere.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case
The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_11_Cyclones.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> Chap_10_Atmosphere.pdf → 84 chunks (107.6s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_11_Cyclones.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case
The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_12_FloodDisasterManagement_p1.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> Chap_11_Cyclones.pdf → 41 chunks (45.1s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_12_FloodDisasterManagement_p1.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> Chap_12_FloodDisasterManagement_p1.pdf → 39 chunks (55.6s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_13_AgricultureDroughtMonitoring.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> Chap_12_FloodDisasterManagement_p2.pdf → 15 chunks (18.4s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_13_AgricultureDroughtMonitoring.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case
The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_14_Landslide.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> Chap_13_AgricultureDroughtMonitoring.pdf → 105 chunks (154.8s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_14_Landslide.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case
The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_15_Earthquake.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> Chap_14_Landslide.pdf → 35 chunks (107.4s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_15_Earthquake.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case
The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_16_ForestFireMonitoring.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> Chap_15_Earthquake.pdf → 50 chunks (138.3s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_16_ForestFireMonitoring.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case
The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_1_Agriculture.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> Chap_16_ForestFireMonitoring.pdf → 101 chunks (342.0s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_1_Agriculture.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case
The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_2_LULC.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> Chap_1_Agriculture.pdf → 72 chunks (226.6s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_2_LULC.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case
The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_3_ForestVegetation_p1.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> Chap_2_LULC.pdf → 122 chunks (244.3s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_3_ForestVegetation_p1.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case
The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_3_ForestVegetation_p2.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> Chap_3_ForestVegetation_p1.pdf → 44 chunks (132.1s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_3_ForestVegetation_p2.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case
The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_3_ForestVegetation_p3.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> Chap_3_ForestVegetation_p2.pdf → 38 chunks (121.4s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_3_ForestVegetation_p3.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case
The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_4_Soils.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> Chap_3_ForestVegetation_p3.pdf → 41 chunks (110.2s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_4_Soils.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case
The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_5_Urban.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> Chap_4_Soils.pdf → 104 chunks (331.5s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_5_Urban.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case
The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_6_Water_RS.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> Chap_5_Urban.pdf → 91 chunks (269.5s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_6_Water_RS.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case
The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_7_Geosciences.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> Chap_6_Water_RS.pdf → 114 chunks (273.6s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_7_Geosciences.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case
The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_8_GroundWater.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> Chap_7_Geosciences.pdf → 160 chunks (443.0s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_8_GroundWater.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case
The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_9_Oceans.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> Chap_8_GroundWater.pdf → 57 chunks (152.1s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\Chap_9_Oceans.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> Chap_9_Oceans.pdf → 158 chunks (365.7s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\OCM2%20NDVI%20Filtered%20Product_for_bhuvan.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> InternationalDayofYoga_Short_bhuvan.pdf → 2 chunks (9.1s)


The PDF <_io.BufferedReader name='..\\..\\data\\pdf\\nrsc-pdfs\\OCM2%20NDVI%20Filtered%20Product_for_bhuvan.pdf'> contains a metadata field indicating that it should not allow text extraction. Ignoring this field and proceeding. Use the check_extractable if you want to raise an error in this case


---> OCM2%20NDVI%20Filtered%20Product_for_bhuvan.pdf → 6 chunks (34.0s)
---> TSC_Bhuvan.pdf → 53 chunks (251.8s)

Total elements: 1563


- *strategy              → HOW to read the PDF (hi_res = smart layout detection)*
- *hi_res_model_name     → WHICH model detects layout (yolox)*
- *infer_table_structure → parse table rows/columns properly*
- *chunking_strategy     → HOW to group text into chunks (by section title)*
- *max_characters        → hard max size of each chunk (1500)*
- *new_after_n_chars     → soft limit, start looking for break point (1200)*
- *combine_text_under    → merge tiny sections so no garbage chunks (300)*
- *multipage_sections    → don't split a section at page boundary*

In [4]:
# Cell 4 — Inspect chunk quality before filtering

# Check length distribution
lengths = [len(el.text) for el in all_elements if hasattr(el, 'text') and el.text]

import statistics
print(f"Total chunks      : {len(all_elements)}")
print(f"Min length        : {min(lengths)} chars")
print(f"Max length        : {max(lengths)} chars")
print(f"Average length    : {int(statistics.mean(lengths))} chars")
print(f"Median length     : {int(statistics.median(lengths))} chars")
print()

# Show chunks under 150 chars — these are garbage candidates
short = [el for el in all_elements if hasattr(el, 'text') and el.text and len(el.text) < 150]
print(f"Chunks under 150 chars: {len(short)}")
print()
print("Sample short chunks:")
for el in short[:10]:
    print(f"  [{el.category}] ({len(el.text)} chars) → {repr(el.text)}")

Total chunks      : 1563
Min length        : 2 chars
Max length        : 1500 chars
Average length    : 974 chars
Median length     : 1094 chars

Chunks under 150 chars: 72

Sample short chunks:
  [CompositeElement] (103 chars) → 'Contact\n\nFor any queries on Bhuvan training kindly email to training@nrsc.gov.in or bhuvan@nrsc.gov.in.'
  [CompositeElement] (42 chars) → 'outlined below (mainly in cases of IMSRA).'
  [CompositeElement] (3 chars) → '272'
  [CompositeElement] (28 chars) → 'Cyclones\n\n11.1. Introduction'
  [CompositeElement] (136 chars) → '11.8. Satellite Technologies\n\nNew technologies have been employed to grasp each of these factors when attempting to make an intensity or'
  [CompositeElement] (136 chars) → 'http://www.mspsofworld.com/hurricane/machanism-of-tropical-cyclone-formation.html\n\nhttp://www.weatherwizkids.com/weather-winds.html\n\n281'
  [CompositeElement] (59 chars) → 'Integration of satellite derived indicators and ground data'
  [CompositeElement] (24 ch

In [5]:
# Cell 5 — Quality Filter

def is_quality_chunk(text):
    if not text:
        return False
    
    text = text.strip()
    
    # Too short — page numbers, headers, fragments
    if len(text) < 150:
        return False
    
    # Mostly URLs — no semantic value
    lines = text.split('\n')
    url_lines = sum(1 for l in lines if l.strip().startswith('http'))
    if url_lines / max(len(lines), 1) > 0.4:
        return False
    
    # Mostly numbers/symbols — garbled OCR or tables without context
    alpha_chars = sum(1 for c in text if c.isalpha())
    if alpha_chars / max(len(text), 1) < 0.4:
        return False
    
    return True

# Apply filter
before = len(all_elements)
clean_elements = [
    el for el in all_elements
    if hasattr(el, 'text') and is_quality_chunk(el.text)
]
after = len(clean_elements)

print(f"Before filtering : {before} chunks")
print(f"After filtering  : {after} chunks")
print(f"Removed          : {before - after} garbage chunks")
print()

# Show what got removed
removed = [
    el for el in all_elements
    if hasattr(el, 'text') and not is_quality_chunk(el.text)
]
print("Sample removed chunks:")
for el in removed[:10]:
    print(f"  ({len(el.text)} chars) → {repr(el.text[:100])}")

Before filtering : 1563 chunks
After filtering  : 1487 chunks
Removed          : 76 garbage chunks

Sample removed chunks:
  (103 chars) → 'Contact\n\nFor any queries on Bhuvan training kindly email to training@nrsc.gov.in or bhuvan@nrsc.gov.'
  (42 chars) → 'outlined below (mainly in cases of IMSRA).'
  (3 chars) → '272'
  (28 chars) → 'Cyclones\n\n11.1. Introduction'
  (136 chars) → '11.8. Satellite Technologies\n\nNew technologies have been employed to grasp each of these factors whe'
  (136 chars) → 'http://www.mspsofworld.com/hurricane/machanism-of-tropical-cyclone-formation.html\n\nhttp://www.weathe'
  (59 chars) → 'Integration of satellite derived indicators and ground data'
  (24 chars) → 'Government of Karnataka.'
  (129 chars) → 'problems such as foreshortening and layover effects associated with SAR data in mountainous areas ha'
  (75 chars) → 'general non-availability of any quantitative technique of model validation.'


In [6]:
# Cell 6 — Convert to LangChain Documents with clean metadata

from langchain_core.documents import Document

def clean_metadata(md: dict) -> dict:
    """Keep only Chroma-compatible types: str, int, float, bool"""
    clean = {}
    for k, v in md.items():
        if isinstance(v, (str, int, float, bool)) or v is None:
            clean[k] = v
        elif isinstance(v, list) and len(v) == 1 and isinstance(v[0], (str, int, float, bool)):
            clean[k] = v[0]
    return clean

# Convert
clean_docs = []
for el in clean_elements:
    md = el.metadata.to_dict()
    md["category"] = el.category
    md.setdefault("source", el.metadata.source)
    
    clean_docs.append(Document(
        page_content=el.text,
        metadata=clean_metadata(md)
    ))

print(f"Total clean documents ready: {len(clean_docs)}")
print()

# Quick sanity check — show one good chunk
sample = clean_docs[10]
print("Sample chunk:")
print(f"  Source   : {sample.metadata.get('source')}")
print(f"  Page     : {sample.metadata.get('page_number')}")
print(f"  Category : {sample.metadata.get('category')}")
print(f"  Length   : {len(sample.page_content)} chars")
print(f"  Preview  : {sample.page_content[:200]}")

Total clean documents ready: 1487

Sample chunk:
  Source   : Bhuvan-NG_2025.pdf
  Page     : 4
  Category : CompositeElement
  Length   : 927 chars
  Preview  : Major Features:

Before After

Get redirected to specialized disaster services portals for detailed information.

Visualization of pre and post disasters with satellite datasets and other spatial data


In [7]:
# Cell 6b — Deduplicate clean_docs

seen_texts = set()
deduped_docs = []

for doc in clean_docs:
    text = doc.page_content.strip()
    if text not in seen_texts:
        seen_texts.add(text)
        deduped_docs.append(doc)

print(f"Before dedup : {len(clean_docs)}")
print(f"After dedup  : {len(deduped_docs)}")
print(f"Removed      : {len(clean_docs) - len(deduped_docs)} duplicates")

# Replace clean_docs
clean_docs = deduped_docs

Before dedup : 1487
After dedup  : 1468
Removed      : 19 duplicates


In [8]:
# Cell 7 — Ingest into 2 Chroma vectorstores

from langchain_ollama import OllamaEmbeddings
from langchain_chroma import Chroma
import time

embedding_models = {
    "nomic" : OllamaEmbeddings(model="nomic-embed-text"),
    "bge"   : OllamaEmbeddings(model="bge-m3"),
}

vectorstores = {}

for name, embeddings in embedding_models.items():
    print(f"\n{'='*50}")
    print(f"Ingesting with {name}...")
    start = time.time()
    
    vs = Chroma.from_documents(
        documents=clean_docs,
        embedding=embeddings,
        persist_directory=f"./chroma_db-{name}",
        collection_name=f"nrsc_docs-{name}"
    )
    
    elapsed = time.time() - start
    mins = int(elapsed // 60)
    secs = int(elapsed % 60)
    
    vectorstores[name] = vs
    print(f"---> {name} done → {vs._collection.count()} chunks indexed in {mins}m {secs}s")

print(f"\n{'='*50}")
print("All 2 vectorstores ready!")


Ingesting with nomic...
---> nomic done → 2936 chunks indexed in 0m 27s

Ingesting with bge...
---> bge done → 2936 chunks indexed in 0m 44s

All 2 vectorstores ready!


nomic-embed-text  → baseline, English focused, 768 dims, 2048 token limit

bge-m3            → multilingual, 1024 dims, 8192 token limit

Dimensions  = output size (how rich the embedding vector is)
              
              nomic=768, bge=1024
              
              MORE dimensions = richer representation

Context limit = how much input text the model can handle
              
              nomic=2048 tokens, bge=8192 tokens
              
              These are INDEPENDENT — not related to each other

In [9]:
for name, vs in vectorstores.items():
    count = vs._collection.count()
    print(f"{name}: {count} chunks in vectorstore")

print(f"\nExpected: {len(clean_docs)} chunks")
print(f"clean_docs length: {len(clean_docs)}")

nomic: 2936 chunks in vectorstore
bge: 2936 chunks in vectorstore

Expected: 1468 chunks
clean_docs length: 1468


In [10]:
# Cell 9 — Setup all retrievers for both vectorstores

from langchain_ollama import OllamaEmbeddings
from langchain_chroma import Chroma
from langchain_community.retrievers import BM25Retriever

# Load vectorstores (if continuing from fresh kernel)
# If vectorstores dict already exists from Cell 7, skip this block
if 'vectorstores' not in dir():
    embedding_models = {
        "nomic" : OllamaEmbeddings(model="nomic-embed-text"),
        "bge"   : OllamaEmbeddings(model="bge-m3"),
    }
    vectorstores = {}
    for name, embeddings in embedding_models.items():
        vectorstores[name] = Chroma(
            persist_directory=f"./chroma_db_{name}",
            embedding_function=embeddings,
            collection_name=f"nrsc_docs_{name}"
        )
        print(f"---> Loaded {name} → {vectorstores[name]._collection.count()} chunks")

# Setup dense retrievers for each vectorstore
retrievers = {}

for name, vs in vectorstores.items():

    retrievers[f"{name}_similarity"] = vs.as_retriever(
        search_type="similarity",
        search_kwargs={"k": 10},
    )

    retrievers[f"{name}_mmr"] = vs.as_retriever(
        search_type="mmr",
        search_kwargs={"k": 10, "fetch_k": 50, "lambda_mult": 0.5},
    )

    retrievers[f"{name}_threshold"] = vs.as_retriever(
        search_type="similarity_score_threshold",
        search_kwargs={"score_threshold": 0.30, "k": 20},
    )

# Sparse retriever — BM25 (embedding independent, only needs text)
bm25_retriever = BM25Retriever.from_documents(clean_docs, bm25_variant="plus")
bm25_retriever.k = 10
retrievers["bm25"] = bm25_retriever

# Print all retrievers
print(f"\nTotal retrievers ready: {len(retrievers)}")
for r in retrievers:
    print(f"  → {r}")


Total retrievers ready: 7
  → nomic_similarity
  → nomic_mmr
  → nomic_threshold
  → bge_similarity
  → bge_mmr
  → bge_threshold
  → bm25


For nomic:
```
Min score  = 0.2872  → best match found across all 15 queries
Max score  = 0.9031  → worst match returned
Mean score = 0.5523  → average distance
25th pct   = 0.4416  → 25% of results are closer than 0.44
50th pct   = 0.5409  → half results are closer than 0.54
75th pct   = 0.6310  → 75% of results are closer than 0.63
90th pct   = 0.8092  → 90% of results are closer than 0.81
```

For bge:
```
Min score  = 0.5334  → best match found
Max score  = 1.0213  → worst match returned
Mean score = 0.7735  → average distance
25th pct   = 0.6752
50th pct   = 0.7759
75th pct   = 0.8429
90th pct   = 0.9506
```

Distance score     → lower = better (0.77 means very close/similar)

Relevance score    → higher = better (0.45 means 45% relevant)

similarity_search_with_score()           → returns distance scores

similarity_search_with_relevance_scores() → returns relevance scores

score_threshold in retriever             → uses RELEVANCE scores

In [11]:
test_query = "What is Bhuvan and when did it start?"

print(f"Query: {test_query}")
print("=" * 60)

for name, retriever in retrievers.items():
    try:
        results = retriever.invoke(test_query)
        top_result = results[0].page_content[:150] if results else "NO RESULTS"
        print(f"\n[{name}]")
        print(f"  Returned : {len(results)} chunks")
        print(f"  Top chunk: {top_result}...")
    except Exception as e:
        print(f"\n[{name}] ERROR: {e}")

Query: What is Bhuvan and when did it start?

[nomic_similarity]
  Returned : 10 chunks
  Top chunk: Formation or Genesis Stage:

Since the nature of Tropical Cyclone (TC) development is continuous, features associated with earliest stages of the TC l...

[nomic_mmr]
  Returned : 10 chunks
  Top chunk: Formation or Genesis Stage:

Since the nature of Tropical Cyclone (TC) development is continuous, features associated with earliest stages of the TC l...

[nomic_threshold]
  Returned : 20 chunks
  Top chunk: Formation or Genesis Stage:

Since the nature of Tropical Cyclone (TC) development is continuous, features associated with earliest stages of the TC l...

[bge_similarity]
  Returned : 10 chunks
  Top chunk: About Bhuvan

Bhuvan, India’s national geospatial platform developed by ISRO and launched in 2009, is a state-of-the-art system integrating diverse ge...

[bge_mmr]
  Returned : 10 chunks
  Top chunk: About Bhuvan

Bhuvan, India’s national geospatial platform developed by ISRO a

In [12]:
# Cell 11 — Find right threshold for each model

test_query = "What is Bhuvan and when did it start?"

print("Checking score distributions...\n")

for name, vs in vectorstores.items():
    results = vs.similarity_search_with_score(test_query, k=15)
    print(f"[{name}]")
    for doc, score in results:
        preview = doc.page_content[:60].replace('\n', ' ')
        print(f"  score={score:.4f} → {preview}...")
    print()

Checking score distributions...

[nomic]
  score=0.8129 → Formation or Genesis Stage:  Since the nature of Tropical Cy...
  score=0.8129 → Formation or Genesis Stage:  Since the nature of Tropical Cy...
  score=0.8332 → Table 13.1: Drought history of India  |  | |  |  Period Drou...
  score=0.8332 → Table 13.1: Drought history of India  |  | |  |  Period Drou...
  score=0.8419 → INSAT system  The Indian National Satellite (INSAT) system w...
  score=0.8419 → INSAT system  The Indian National Satellite (INSAT) system w...
  score=0.8623 → 10.4. Present and Future Missions  10.4.1. INSAT Series  INS...
  score=0.8623 → 10.4. Present and Future Missions  10.4.1. INSAT Series  INS...
  score=0.8705 → EDUSAT  Configured for audio-visual medium employing digital...
  score=0.8705 → EDUSAT  Configured for audio-visual medium employing digital...
  score=0.8819 → The DORIS (Doppler Orbitography and Radiopositioning Integra...
  score=0.8819 → The DORIS (Doppler Orbitography and Radiopositionin

```
bge   → Bhuvan chunk is at distance 0.77  → rank 1
nomic → Bhuvan chunk is at distance 0.90  → rank 10
```

nomic pushed the correct answer 10x further away than bge did. That's the entire problem in one number.

And this is why our old results were 71% hit rate:

Those were all nomic results. bge will likely give you 85%+ because it actually understands our content.

In [13]:
# Check for duplicates in bge vectorstore
bge_count = vectorstores["bge"]._collection.count()
print(f"bge vectorstore count : {bge_count}")
print(f"clean_docs count      : {len(clean_docs)}")

# Check duplicates in clean_docs itself
unique_texts = set(d.page_content for d in clean_docs)
print(f"Unique texts in clean_docs: {len(unique_texts)}")

bge vectorstore count : 2936
clean_docs count      : 1468
Unique texts in clean_docs: 1468


In [14]:
# Cell 12 — Find threshold properly with multiple queries

test_queries = [
    "What is Bhuvan and when did it start?",
    "What is remote sensing?",
    "How are floods monitored using satellites?",
    "What is NDVI and how is it calculated?",
    "What are the types of soil in India?",
    "How does cyclone formation happen?",
    "What is land use land cover mapping?",
    "How is groundwater detected using remote sensing?",
    "What is the role of ISRO in disaster management?",
    "What are the applications of SAR data?",
    "How is drought monitored using satellite data?",
    "What is the difference between active and passive sensors?",
    "What is photosynthesis in remote sensing context?",
    "How are landslides detected from satellite imagery?",
    "What is ocean color remote sensing?",
]

import numpy as np

for vs_name, vs in vectorstores.items():
    all_scores = []
    
    for query in test_queries:
        results = vs.similarity_search_with_score(query, k=10)
        scores = [score for _, score in results]
        all_scores.extend(scores)
    
    all_scores = sorted(all_scores)
    
    print(f"\n[{vs_name}] across {len(test_queries)} queries:")
    print(f"  Min score  : {min(all_scores):.4f}")
    print(f"  Max score  : {max(all_scores):.4f}")
    print(f"  Mean score : {np.mean(all_scores):.4f}")
    print(f"  25th pct   : {np.percentile(all_scores, 25):.4f}")
    print(f"  50th pct   : {np.percentile(all_scores, 50):.4f}")
    print(f"  75th pct   : {np.percentile(all_scores, 75):.4f}")
    print(f"  90th pct   : {np.percentile(all_scores, 90):.4f}")


[nomic] across 15 queries:
  Min score  : 0.2872
  Max score  : 0.8705
  Mean score : 0.5159
  25th pct   : 0.4078
  50th pct   : 0.4815
  75th pct   : 0.6081
  90th pct   : 0.8020

[bge] across 15 queries:
  Min score  : 0.5334
  Max score  : 1.0101
  Mean score : 0.7331
  25th pct   : 0.6366
  50th pct   : 0.7199
  75th pct   : 0.8214
  90th pct   : 0.9000


```
               75th pct   90th pct   Recommended threshold
nomic          0.6310     0.8092    → 0.65
bge            0.8429     0.9506    → 0.87
```

nomic threshold = 0.65

→ top 75% of results pass through

→ bottom 25% (likely noise) gets blocked

bge threshold = 0.87  

→ top 75% of results pass through

→ bottom 25% (likely noise) gets blocked

In [15]:
# Cell 13 — Final threshold values based on 15-query analysis

retrievers["nomic_threshold"] = vectorstores["nomic"].as_retriever(
    search_type="similarity_score_threshold",
    search_kwargs={"score_threshold": 0.65, "k": 20},
)

retrievers["bge_threshold"] = vectorstores["bge"].as_retriever(
    search_type="similarity_score_threshold",
    search_kwargs={"score_threshold": 0.87, "k": 20},
)

print("---> Threshold retrievers updated based on 15-query analysis")
print(f"nomic threshold → 0.65  (75th percentile = 0.6310)")
print(f"bge threshold   → 0.87  (75th percentile = 0.8429)")

---> Threshold retrievers updated based on 15-query analysis
nomic threshold → 0.65  (75th percentile = 0.6310)
bge threshold   → 0.87  (75th percentile = 0.8429)


*Threshold values were determined empirically by analyzing the score distributions of each embedding model on a sample query set, rather than using arbitrary fixed values.*

*"Threshold was determined by analyzing score distributions across 15 diverse domain queries rather than a single query, ensuring statistical validity."*



In [16]:
# Quick check
test_query = "What is Bhuvan and when did it start?"

for name, retriever in retrievers.items():
    try:
        results = retriever.invoke(test_query)
        top = results[0].page_content[:80].replace('\n',' ') if results else "NO RESULTS"
        print(f"[{name}] {len(results)} results → {top}...")
    except Exception as e:
        print(f"[{name}] ❌ {e}")

[nomic_similarity] 10 results → Formation or Genesis Stage:  Since the nature of Tropical Cyclone (TC) developme...
[nomic_mmr] 10 results → Formation or Genesis Stage:  Since the nature of Tropical Cyclone (TC) developme...


No relevant docs were retrieved using the relevance score threshold 0.65


[nomic_threshold] 0 results → NO RESULTS...
[bge_similarity] 10 results → About Bhuvan  Bhuvan, India’s national geospatial platform developed by ISRO and...
[bge_mmr] 10 results → About Bhuvan  Bhuvan, India’s national geospatial platform developed by ISRO and...


No relevant docs were retrieved using the relevance score threshold 0.87


[bge_threshold] 0 results → NO RESULTS...
[bm25] 10 results → Bhuvan NextGen Inception  Natural resources information for 10 themes.  Bhuvan N...


In [17]:
# Cell 13 — Debug threshold issue

# Check what similarity_search_with_relevance_scores returns
# This is different from similarity_search_with_score

test_query = "What is Bhuvan and when did it start?"

for vs_name, vs in vectorstores.items():
    print(f"\n[{vs_name}] relevance scores (higher = better):")
    results = vs.similarity_search_with_relevance_scores(test_query, k=10)
    for doc, score in results:
        preview = doc.page_content[:60].replace('\n', ' ')
        print(f"  relevance={score:.4f} → {preview}...")


[nomic] relevance scores (higher = better):
  relevance=0.4252 → Formation or Genesis Stage:  Since the nature of Tropical Cy...
  relevance=0.4252 → Formation or Genesis Stage:  Since the nature of Tropical Cy...
  relevance=0.4109 → Table 13.1: Drought history of India  |  | |  |  Period Drou...
  relevance=0.4109 → Table 13.1: Drought history of India  |  | |  |  Period Drou...
  relevance=0.4047 → INSAT system  The Indian National Satellite (INSAT) system w...
  relevance=0.4047 → INSAT system  The Indian National Satellite (INSAT) system w...
  relevance=0.3903 → 10.4. Present and Future Missions  10.4.1. INSAT Series  INS...
  relevance=0.3903 → 10.4. Present and Future Missions  10.4.1. INSAT Series  INS...
  relevance=0.3844 → EDUSAT  Configured for audio-visual medium employing digital...
  relevance=0.3844 → EDUSAT  Configured for audio-visual medium employing digital...

[bge] relevance scores (higher = better):
  relevance=0.4510 → About Bhuvan  Bhuvan, India’s national ge

In [18]:
# Cell 13 — Debug threshold issue

# Check what similarity_search_with_relevance_scores returns
# This is different from similarity_search_with_score

test_query = "What is Bhuvan and when did it start?"

for vs_name, vs in vectorstores.items():
    print(f"\n[{vs_name}] relevance scores (higher = better):")
    results = vs.similarity_search_with_relevance_scores(test_query, k=10)
    for doc, score in results:
        preview = doc.page_content[:60].replace('\n', ' ')
        print(f"  relevance={score:.4f} → {preview}...")


[nomic] relevance scores (higher = better):
  relevance=0.4252 → Formation or Genesis Stage:  Since the nature of Tropical Cy...
  relevance=0.4252 → Formation or Genesis Stage:  Since the nature of Tropical Cy...
  relevance=0.4109 → Table 13.1: Drought history of India  |  | |  |  Period Drou...
  relevance=0.4109 → Table 13.1: Drought history of India  |  | |  |  Period Drou...
  relevance=0.4047 → INSAT system  The Indian National Satellite (INSAT) system w...
  relevance=0.4047 → INSAT system  The Indian National Satellite (INSAT) system w...
  relevance=0.3903 → 10.4. Present and Future Missions  10.4.1. INSAT Series  INS...
  relevance=0.3903 → 10.4. Present and Future Missions  10.4.1. INSAT Series  INS...
  relevance=0.3844 → EDUSAT  Configured for audio-visual medium employing digital...
  relevance=0.3844 → EDUSAT  Configured for audio-visual medium employing digital...

[bge] relevance scores (higher = better):
  relevance=0.4510 → About Bhuvan  Bhuvan, India’s national ge

Distance scores  → 0.28 to 1.09  (lower = better)

Relevance scores → 0.0  to 1.0   (higher = better)

nomic best relevance → 0.4252

bge   best relevance → 0.4510

This explains why threshold was returning 0 results.
We set threshold = 0.65 and 0.87 based on distance scores. But the threshold retriever uses relevance scores internally. Your best results only reach 0.45 relevance — so nothing ever passed 0.65 or 0.87.

In [19]:
# Cell 12 — Correct threshold based on RELEVANCE scores

retrievers["nomic_threshold"] = vectorstores["nomic"].as_retriever(
    search_type="similarity_score_threshold",
    search_kwargs={"score_threshold": 0.38, "k": 20},
)

retrievers["bge_threshold"] = vectorstores["bge"].as_retriever(
    search_type="similarity_score_threshold",
    search_kwargs={"score_threshold": 0.40, "k": 20},
)

print("---> Threshold retrievers updated based on RELEVANCE scores")
print(f"nomic threshold → 0.38")
print(f"bge threshold   → 0.40")

---> Threshold retrievers updated based on RELEVANCE scores
nomic threshold → 0.38
bge threshold   → 0.40


In [20]:
test_query = "What is Bhuvan and when did it start?"

for name, retriever in retrievers.items():
    try:
        results = retriever.invoke(test_query)
        top = results[0].page_content[:80].replace('\n',' ') if results else "NO RESULTS"
        print(f"[{name}] {len(results)} results → {top}...")
    except Exception as e:
        print(f"[{name}] {e}")

[nomic_similarity] 10 results → Formation or Genesis Stage:  Since the nature of Tropical Cyclone (TC) developme...
[nomic_mmr] 10 results → Formation or Genesis Stage:  Since the nature of Tropical Cyclone (TC) developme...
[nomic_threshold] 10 results → Formation or Genesis Stage:  Since the nature of Tropical Cyclone (TC) developme...
[bge_similarity] 10 results → About Bhuvan  Bhuvan, India’s national geospatial platform developed by ISRO and...
[bge_mmr] 10 results → About Bhuvan  Bhuvan, India’s national geospatial platform developed by ISRO and...
[bge_threshold] 4 results → About Bhuvan  Bhuvan, India’s national geospatial platform developed by ISRO and...
[bm25] 10 results → Bhuvan NextGen Inception  Natural resources information for 10 themes.  Bhuvan N...


*"Initial threshold values were incorrectly derived from distance scores. Upon discovering that LangChain's score_threshold parameter uses relevance scores internally, thresholds were recalibrated by directly inspecting relevance score distributions. Final values: nomic=0.38, bge=0.40, determined by identifying natural gaps in the relevance score distributions."*

*nomic-embed-text showed compressed relevance score distributions (range: 0.36–0.42) with minimal separation between relevant and irrelevant chunks, making threshold-based filtering unreliable. bge-m3 showed clearer score separation, enabling more effective threshold filtering*

In [21]:
# Cell — Eval Dataset Generation (v2)

import pandas as pd
import time
from langchain_ollama import OllamaLLM
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.output_parsers import StrOutputParser

# LLM
llm = OllamaLLM(
    model="llama3.2:3b",
    temperature=0.0,
)

# Prompt
prompt = ChatPromptTemplate.from_template("""
You are an expert in remote sensing, geospatial science, 
and earth observation technologies.

Given the passage below, generate exactly 2 specific factual 
questions that a domain expert or student would genuinely ask 
about this topic.

STRICT RULES:
- Questions must be answerable ONLY from this passage
- ONLY use information explicitly present in the passage
- Do NOT use outside knowledge or add extra information
- Do NOT ask about author names, publication years, journal names
- Do NOT ask about figure numbers or table numbers
- Do NOT reference "the passage" or "the text" in questions
- Do NOT hallucinate facts not present in the passage
- Focus ONLY on: concepts, definitions, processes,
  findings, numbers, applications, techniques
- Each question must be self-contained and clear
- Both questions must be different from each other

Return ONLY 2 questions, one per line.
No numbering. No extra text. No explanation.

Passage:
{text}
""")

chain = prompt | llm | StrOutputParser()

# Resume capability
try:
    eval_df = pd.read_csv("eval_v2/eval_dataset_v2.csv")
    done_ids = set(eval_df["chunk_id"].unique())
    print(f"Resuming — {len(done_ids)} chunks already done")
    print(f"Questions so far — {len(eval_df)}")
except FileNotFoundError:
    eval_df = pd.DataFrame()
    done_ids = set()
    print("Starting fresh")

# Build registry from clean_docs
registry = []
for i, doc in enumerate(clean_docs):
    chunk_id = f"chunk_{str(i+1).zfill(4)}"
    registry.append({
        "chunk_id"   : chunk_id,
        "text"       : doc.page_content,
        "source_file": doc.metadata.get("source", ""),
        "page_number": doc.metadata.get("page_number", ""),
    })

registry_df = pd.DataFrame(registry)
remaining = registry_df[
    ~registry_df["chunk_id"].isin(done_ids)
].reset_index(drop=True)

print(f"\nTotal chunks    : {len(registry_df)}")
print(f"Already done    : {len(done_ids)}")
print(f"Remaining       : {len(remaining)}")
print(f"Estimated time  : ~{len(remaining) * 4 // 60} mins")
print(f"Total questions : ~{len(remaining) * 2} questions")
print("\nStarting generation...")

Starting fresh

Total chunks    : 1468
Already done    : 0
Remaining       : 1468
Estimated time  : ~97 mins
Total questions : ~2936 questions

Starting generation...


In [22]:
# Generation loop for questions

records = []
checkpoint_every = 50

for i, row in remaining.iterrows():
    try:
        response = chain.invoke({"text": row["text"][:1200]})
        
        lines = [
            q.strip() for q in response.strip().split("\n")
            if q.strip() and len(q.strip()) > 20
        ]
        questions = lines[:2]

        if len(questions) < 2:
            print(f"[WARN] {row['chunk_id']} → only {len(questions)} question(s), skipping")
            continue

        for q in questions:
            records.append({
                "question"          : q,
                "chunk_id"          : row["chunk_id"],
                "source_file"       : row["source_file"],
                "page_number"       : row["page_number"],
                "ground_truth_text" : row["text"],
            })

        # Checkpoint every 50 chunks
        if (i + 1) % checkpoint_every == 0:
            checkpoint_df = pd.DataFrame(records)
            combined = pd.concat([eval_df, checkpoint_df], ignore_index=True)
            combined.to_csv("eval_dataset_v2.csv", index=False)
            eval_df = combined
            records = []
            print(f"[{i+1}/{len(remaining)}] Saved — Total questions: {len(eval_df)}")

    except Exception as e:
        print(f"[ERROR] {row['chunk_id']}: {e}")
        continue

# Final save
if records:
    checkpoint_df = pd.DataFrame(records)
    combined = pd.concat([eval_df, checkpoint_df], ignore_index=True)
    combined.to_csv("eval_dataset_v2.csv", index=False)
    print(f"\n---> Done! Total questions saved: {len(combined)}")
else:
    print(f"\n---> Done! Total questions saved: {len(eval_df)}")

[50/1468] Saved — Total questions: 100
[100/1468] Saved — Total questions: 200
[150/1468] Saved — Total questions: 300
[200/1468] Saved — Total questions: 400
[250/1468] Saved — Total questions: 500
[300/1468] Saved — Total questions: 600
[350/1468] Saved — Total questions: 700
[400/1468] Saved — Total questions: 800
[450/1468] Saved — Total questions: 900
[500/1468] Saved — Total questions: 1000
[550/1468] Saved — Total questions: 1100
[600/1468] Saved — Total questions: 1200
[650/1468] Saved — Total questions: 1300
[700/1468] Saved — Total questions: 1400
[750/1468] Saved — Total questions: 1500
[800/1468] Saved — Total questions: 1600
[850/1468] Saved — Total questions: 1700
[900/1468] Saved — Total questions: 1800
[950/1468] Saved — Total questions: 1900
[WARN] chunk_1000 → only 1 question(s), skipping
[1050/1468] Saved — Total questions: 2098
[1100/1468] Saved — Total questions: 2198
[1150/1468] Saved — Total questions: 2298
[1200/1468] Saved — Total questions: 2398
[1250/1468] Sa

In [23]:
# Cell — Verify eval dataset quality

import pandas as pd

eval_df = pd.read_csv("eval_dataset_v2.csv")

print(f"Total questions  : {len(eval_df)}")
print(f"Total chunks     : {eval_df['chunk_id'].nunique()}")
print(f"Source files     : {eval_df['source_file'].nunique()}")
print()

# Show sample questions
print("=== Sample questions ===")
for i, row in eval_df.sample(10, random_state=42).iterrows():
    print(f"Q: {row['question']}")
    print(f"Source: {row['source_file']} | Page: {row['page_number']}")
    print(f"Ground truth preview: {row['ground_truth_text'][:100]}...")
    print()

# Check per source file distribution
print("=== Questions per source file ===")
print(eval_df['source_file'].value_counts())

Total questions  : 2934
Total chunks     : 1467
Source files     : 26

=== Sample questions ===
Q: What is the primary mechanism by which kames are formed?
Source: Chap_7_Geosciences.pdf | Page: 36
Ground truth preview: Kame Terrace: Kames are mounds of poorly sorted sand and gravel deposited from running water in clos...

Q: What are the estimated fish catch ranges in hall and other areas along the north Andhra and Orissa coast?
Source: Chap_9_Oceans.pdf | Page: 10
Ground truth preview: S a FE @ 86

Figure 9.4: Synergy of SST and Chlorophyll derived from NOAA and OCM respectively

2002...

Q: What is the primary purpose of the AVISO User Handbook Ssalto/Duacs?
Source: TSC_Bhuvan.pdf | Page: 25
Ground truth preview: 6. References

AVISO User Handbook Ssalto/Duacs: M(SLA) and M(ADT) Near-Real Time and Delayed- Time,...

Q: How does the movement of underground water relate to faults in terms of forming springs and ponds?
Source: Chap_15_Earthquake.pdf | Page: 5
Ground truth preview: Most

In [24]:
# Remove questions from reference chunks
import pandas as pd

eval_df = pd.read_csv("eval_dataset_v2.csv")

# Filter out chunks that are clearly reference sections
reference_keywords = [
    "References\n", 
    "User Handbook",
    "et al.,",
    "Journal of",
    "Vol.",
    "pp.",
]

def is_reference_chunk(text):
    count = sum(1 for kw in reference_keywords if kw in text)
    return count >= 2

before = len(eval_df)
eval_df = eval_df[
    ~eval_df["ground_truth_text"].apply(is_reference_chunk)
].reset_index(drop=True)
after = len(eval_df)

print(f"Before : {before} questions")
print(f"After  : {after} questions")
print(f"Removed: {before - after} bibliography questions")

eval_df.to_csv("eval_dataset_v2.csv", index=False)
print("---> Saved clean eval dataset")

Before : 2934 questions
After  : 2894 questions
Removed: 40 bibliography questions
---> Saved clean eval dataset


In [25]:
# Also remove questions where ground truth is garbled OCR
def is_garbled(text):
    words = text.split()
    # Too many single character words = garbled
    single_chars = sum(1 for w in words if len(w) == 1)
    return single_chars / max(len(words), 1) > 0.3

before = len(eval_df)
eval_df = eval_df[
    ~eval_df["ground_truth_text"].apply(is_garbled)
].reset_index(drop=True)
after = len(eval_df)

print(f"Before : {before}")
print(f"After  : {after}")
print(f"Removed: {before - after} garbled chunks")

eval_df.to_csv("eval_dataset_v2.csv", index=False)
print("---> Saved final clean eval dataset")

Before : 2894
After  : 2878
Removed: 16 garbled chunks
---> Saved final clean eval dataset


In [26]:
# Check total after both filters
import pandas as pd

eval_df = pd.read_csv("eval_dataset_v2.csv")
print(f"Final clean eval dataset: {len(eval_df)} questions")
print(f"Total chunks covered    : {eval_df['chunk_id'].nunique()}")
print(f"Started with            : 2934 questions")
print(f"Total removed           : {2934 - len(eval_df)} questions")

Final clean eval dataset: 2878 questions
Total chunks covered    : 1439
Started with            : 2934 questions
Total removed           : 56 questions


In [1]:
import pickle

with open("clean_docs.pkl", "rb") as f:
    clean_docs = pickle.load(f)

print(f"Loaded {len(clean_docs)} docs")
print("Metadata keys:", list(clean_docs[0].metadata.keys()))

Loaded 1487 docs
Metadata keys: ['filetype', 'last_modified', 'page_number', 'file_directory', 'filename', 'source', 'category']


In [2]:
# Remove orig_elements from all docs — it's useless and huge
from langchain_core.documents import Document
import pickle

clean_docs = [
    Document(
        page_content=doc.page_content,
        metadata={
            k: v for k, v in doc.metadata.items()
            if k != "orig_elements"  # remove this huge blob
        }
    )
    for doc in clean_docs
]

# Save
with open("clean_docs.pkl", "wb") as f:
    pickle.dump(clean_docs, f)

print(f"---> Cleaned {len(clean_docs)} docs")
print("Metadata keys now:", list(clean_docs[0].metadata.keys()))

---> Cleaned 1487 docs
Metadata keys now: ['filetype', 'last_modified', 'page_number', 'file_directory', 'filename', 'source', 'category']


In [2]:
import shutil, os, time
from langchain_ollama import OllamaEmbeddings
from langchain_chroma import Chroma

embedding_models = {
    "nomic" : OllamaEmbeddings(model="nomic-embed-text"),
    "bge"   : OllamaEmbeddings(model="bge-m3"),
}

vectorstores = {}
for name, embeddings in embedding_models.items():
    print(f"Ingesting {name}...")
    start = time.time()
    vs = Chroma.from_documents(
        documents=clean_docs,
        embedding=embeddings,
        persist_directory=f"./chroma_db_{name}",
        collection_name=f"nrsc_docs_{name}"
    )
    elapsed = time.time() - start
    vectorstores[name] = vs
    print(f"{name} --> {vs._collection.count()} chunks in {int(elapsed//60)}m {int(elapsed%60)}s")

# Rebuild retrievers
retrievers = {}
for name, vs in vectorstores.items():
    retrievers[f"{name}_similarity"] = vs.as_retriever(
        search_type="similarity",
        search_kwargs={"k": 10},
    )
    retrievers[f"{name}_mmr"] = vs.as_retriever(
        search_type="mmr",
        search_kwargs={"k": 10, "fetch_k": 50, "lambda_mult": 0.5},
    )
    retrievers[f"{name}_threshold"] = vs.as_retriever(
        search_type="similarity_score_threshold",
        search_kwargs={
            "score_threshold": 0.38 if name == "nomic" else 0.40,
            "k": 20
        },
    )

from langchain_community.retrievers import BM25Retriever
bm25_retriever = BM25Retriever.from_documents(clean_docs, bm25_variant="plus")
bm25_retriever.k = 10
retrievers["bm25"] = bm25_retriever

print(f"\nTotal retrievers: {len(retrievers)}")
for i, r in enumerate(retrievers):
    print(f"  {i+1} {r}")

Ingesting nomic...
nomic --> 1487 chunks in 0m 18s
Ingesting bge...
bge --> 1487 chunks in 0m 41s

Total retrievers: 7
  1 nomic_similarity
  2 nomic_mmr
  3 nomic_threshold
  4 bge_similarity
  5 bge_mmr
  6 bge_threshold
  7 bm25


In [6]:
# Test bge-m3 embedding on problematic questions

from langchain_ollama import OllamaEmbeddings

bge_test = OllamaEmbeddings(model="bge-m3")

test_questions = [
    "What is the primary application of Bhuvan NextGen, specifically in terms of its target users?",
    "How many campuses does NRSC operate from",
    "What percentage of the dry atmosphere is",
    "What percentage increase in carbon dioxide",
    "How much has the amount of methane produced",
    "What percentage of nitrogen is stored in",
    "What types of clouds can be determined b",
    "What is remote sensing?",  # control question - should work fine
]

for q in test_questions:
    try:
        embedding = bge_test.embed_query(q)
        # Check for NaN in the embedding
        nan_count = sum(1 for v in embedding if v != v)  # NaN != NaN is True
        if nan_count > 0:
            print(f"NaN DETECTED ({nan_count} values) --> {q[:60]}")
        else:
            print(f"OK ({len(embedding)} dims) --> {q[:60]}")
    except Exception as e:
        print(f"ERROR --> {q[:60]}")
        print(f"  {e}")

ERROR --> What is the primary application of Bhuvan NextGen, specifica
  failed to encode response: json: unsupported value: NaN (status code: 500)
OK (1024 dims) --> How many campuses does NRSC operate from
OK (1024 dims) --> What percentage of the dry atmosphere is
OK (1024 dims) --> What percentage increase in carbon dioxide
OK (1024 dims) --> How much has the amount of methane produced
OK (1024 dims) --> What percentage of nitrogen is stored in
OK (1024 dims) --> What types of clouds can be determined b
OK (1024 dims) --> What is remote sensing?


In [7]:
# Diagnose the exact problematic question

problem_q = "What is the primary application of Bhuvan NextGen, specifically in terms of its target users?"

# Test 1 - is it the length?
print(f"Length: {len(problem_q)} chars")

# Test 2 - test shorter version
try:
    e = bge_test.embed_query("What is the primary application of Bhuvan NextGen")
    print("Shorter version: OK")
except:
    print("Shorter version: FAILED")

# Test 3 - is it the word "NextGen"?
try:
    e = bge_test.embed_query("What is NextGen?")
    print("NextGen alone: OK")
except:
    print("NextGen alone: FAILED")

# Test 4 - is it the Hindi content in the CHUNK being retrieved?
# The chunk for this question contains Hindi text
problem_chunk = "भुवन अगलीपीढी Bhuvan NextGen https://bhuvanmaps.nrsc.gov.in Standard Mode"
try:
    e = bge_test.embed_documents([problem_chunk])
    nan_count = sum(1 for v in e[0] if v != v)
    print(f"Hindi chunk embedding: OK, NaN count={nan_count}")
except Exception as ex:
    print(f"Hindi chunk embedding: FAILED -- {ex}")

Length: 93 chars
Shorter version: OK
NextGen alone: OK
Hindi chunk embedding: FAILED -- failed to encode response: json: unsupported value: NaN (status code: 500)


In [8]:
import pandas as pd

eval_df = pd.read_csv("eval_dataset_v2.csv")

def has_problematic_hindi(text):
    devanagari = sum(1 for c in str(text) if '\u0900' <= c <= '\u097F')
    total = len(str(text))
    return devanagari / max(total, 1) > 0.15

before = len(eval_df)
eval_df = eval_df[
    ~eval_df["ground_truth_text"].apply(has_problematic_hindi)
].reset_index(drop=True)
after = len(eval_df)

print(f"Before  : {before}")
print(f"After   : {after}")
print(f"Removed : {before - after} Hindi-heavy questions")

eval_df.to_csv("eval_dataset_v2.csv", index=False)
print("Saved clean eval dataset")

Before  : 2878
After   : 2878
Removed : 0 Hindi-heavy questions
Saved clean eval dataset


In [10]:
# Find which chunks cause NaN when embedded by bge

from langchain_ollama import OllamaEmbeddings
import pandas as pd

bge_test = OllamaEmbeddings(model="bge-m3")

problematic_chunks = []

for i, doc in enumerate(clean_docs):
    try:
        embedding = bge_test.embed_documents([doc.page_content])
        nan_count = sum(1 for v in embedding[0] if v != v)
        if nan_count > 0:
            problematic_chunks.append(i)
            print(f"NaN in chunk {i}: {doc.page_content[:80]}")
    except Exception as e:
        problematic_chunks.append(i)
        print(f"ERROR chunk {i}: {doc.page_content[:80]}")
        print(f"  --> {e}")

print(f"\nTotal problematic chunks: {len(problematic_chunks)}")
print(f"Indices: {problematic_chunks}")

KeyboardInterrupt: 

In [11]:
# Test the exact failing questions directly

bge_test = OllamaEmbeddings(model="bge-m3")

failing_questions = [
    "What types of clouds can be determined b",
    "What is the primary purpose of combining",
    "How does reflected solar radiation compa",
]

for q in failing_questions:
    try:
        e = bge_test.embed_query(q)
        print(f"OK --> {q}")
    except Exception as ex:
        print(f"FAILED --> {q}")
        print(f"  {ex}")

OK --> What types of clouds can be determined b
OK --> What is the primary purpose of combining
OK --> How does reflected solar radiation compa


In [12]:
import os
if os.path.exists("eval_results_v2.csv"):
    os.remove("eval_results_v2.csv")
    print("Deleted partial results - starting fresh")

Deleted partial results - starting fresh


In [15]:
import os
if os.path.exists("eval_results_v2.csv"):
    os.remove("eval_results_v2.csv")
    print("Deleted partial results")
else:
    print("No existing results file found")

No existing results file found


In [16]:
import pandas as pd
eval_df = pd.read_csv("eval_dataset_v2.csv")

problem = eval_df[eval_df["question"].str.contains("primary application of Bhuva", na=False)]
print(f"Found {len(problem)} matching rows")
if len(problem) > 0:
    print(problem[["question", "chunk_id"]].to_string())

Found 1 matching rows
                                                                                        question    chunk_id
6  What is the primary application of Bhuvan NextGen, specifically in terms of its target users?  chunk_0004


In [17]:
import pandas as pd

eval_df = pd.read_csv("eval_dataset_v2.csv")

# Test each question with bge before evaluation
from langchain_ollama import OllamaEmbeddings

bge_test = OllamaEmbeddings(model="bge-m3")

problematic_indices = []

print("Testing all questions with bge-m3...")
for i, row in eval_df.iterrows():
    try:
        bge_test.embed_query(row["question"])
    except:
        problematic_indices.append(i)
        print(f"FAILED --> [{i}] {row['question'][:80]}")

print(f"\nTotal problematic questions: {len(problematic_indices)}")

# Remove them
eval_df = eval_df.drop(index=problematic_indices).reset_index(drop=True)
eval_df.to_csv("eval_dataset_v2.csv", index=False)
print(f"Remaining questions: {len(eval_df)}")

FAILED --> [1553] How many field observations are initially required to confirm a hypothesis about
FAILED --> [1556] What are the three soil-forming factors that can be analyzed by the Geo-Pedologi
FAILED --> [1573] How many types of soil profiles are mentioned as being studied along with morpho
FAILED --> [1583] Can non-normal distributions be considered by classification algorithms that do 
FAILED --> [1595] How did stereoscopy from IRS-1D PAN stereo contribute to the delineation of ravi
FAILED --> [1609] How does frost heaving affect the germination and root growth of crops in these 


: 

: 

In [1]:
# Test sending question one by one vs batch

from langchain_ollama import OllamaEmbeddings

bge_test = OllamaEmbeddings(model="bge-m3")

problem_q = "What is the primary application of Bhuvan NextGen, specifically in terms of its target users?"

# Test 1 - single query
try:
    e = bge_test.embed_query(problem_q)
    print(f"embed_query: OK")
except Exception as ex:
    print(f"embed_query: FAILED -- {ex}")

# Test 2 - as document (different code path)
try:
    e = bge_test.embed_documents([problem_q])
    print(f"embed_documents single: OK")
except Exception as ex:
    print(f"embed_documents single: FAILED -- {ex}")

# Test 3 - batch of 2
try:
    e = bge_test.embed_documents([problem_q, "What is remote sensing?"])
    print(f"embed_documents batch: OK")
except Exception as ex:
    print(f"embed_documents batch: FAILED -- {ex}")

e:\The Leo Programmer\Internships\NRSC\Assignments\testing-models-huge-dataset\.venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


embed_query: FAILED -- failed to encode response: json: unsupported value: NaN (status code: 500)
embed_documents single: FAILED -- failed to encode response: json: unsupported value: NaN (status code: 500)
embed_documents batch: FAILED -- failed to encode response: json: unsupported value: NaN (status code: 500)


In [14]:
import pandas as pd
from tqdm import tqdm

eval_df = pd.read_csv("eval_dataset_v2.csv")
print(f"Total questions : {len(eval_df)}")
print(f"Total retrievers: {len(retrievers)}")
print(f"Total API calls : {len(eval_df) * len(retrievers)}")

try:
    results_df = pd.read_csv("eval_results_v2.csv")
    done_questions = set(zip(results_df["question"], results_df["retriever"]))
    print(f"Resuming -- {len(results_df)} results already saved")
except FileNotFoundError:
    results_df = pd.DataFrame()
    done_questions = set()
    print("Starting fresh")

print("\nStarting evaluation...")
print("=" * 50)

records = []
checkpoint_every = 100

for i, row in tqdm(eval_df.iterrows(), total=len(eval_df)):
    question     = row["question"]
    ground_truth = row["ground_truth_text"]
    chunk_id     = row["chunk_id"]

    for retriever_name, retriever in retrievers.items():

        if (question, retriever_name) in done_questions:
            continue

        try:
            docs = retriever.invoke(question)
            retrieved_texts = [doc.page_content for doc in docs]

            hit        = 0
            rank       = None
            n_returned = len(retrieved_texts)

            for r, text in enumerate(retrieved_texts):
                if text.strip() == ground_truth.strip():
                    hit  = 1
                    rank = r + 1
                    break

            records.append({
                "question"   : question,
                "chunk_id"   : chunk_id,
                "retriever"  : retriever_name,
                "hit"        : hit,
                "rank"       : rank,
                "n_returned" : n_returned,
            })

        except Exception as e:
            print(f"\n[ERROR] {retriever_name} | {question[:40]}: {e}")
            # Record as miss so question is not skipped entirely
            records.append({
                "question"   : question,
                "chunk_id"   : chunk_id,
                "retriever"  : retriever_name,
                "hit"        : 0,
                "rank"       : None,
                "n_returned" : 0,
            })
            continue

    if (i + 1) % checkpoint_every == 0:
        checkpoint_df = pd.DataFrame(records)
        combined = pd.concat([results_df, checkpoint_df], ignore_index=True)
        combined.to_csv("eval_results_v2.csv", index=False)
        results_df = combined
        records = []
        print(f"\n[{i+1}/{len(eval_df)}] Saved -- {len(results_df)} results")

if records:
    checkpoint_df = pd.DataFrame(records)
    combined = pd.concat([results_df, checkpoint_df], ignore_index=True)
    combined.to_csv("eval_results_v2.csv", index=False)
    print(f"\nDone! Total results saved: {len(combined)}")
else:
    print(f"\nDone! Total results saved: {len(results_df)}")

Total questions : 2878
Total retrievers: 7
Total API calls : 20146
Starting fresh

Starting evaluation...


  0%|          | 6/2878 [00:13<1:46:37,  2.23s/it]


[ERROR] bge_similarity | What is the primary application of Bhuva: failed to encode response: json: unsupported value: NaN (status code: 500)

[ERROR] bge_mmr | What is the primary application of Bhuva: failed to encode response: json: unsupported value: NaN (status code: 500)


  0%|          | 7/2878 [00:15<1:44:49,  2.19s/it]


[ERROR] bge_threshold | What is the primary application of Bhuva: failed to encode response: json: unsupported value: NaN (status code: 500)


  1%|          | 15/2878 [00:34<1:50:07,  2.31s/it]


KeyboardInterrupt: 

You're going in circles! Let me recap the full chain:

```
bge-m3 via Ollama → NaN values → 500 error
         ↓
    Why NaN?
         ↓
bge-m3 is a multi-vector model (dense + sparse + colbert)
         ↓
Ollama uses llama.cpp which only understands simple BERT-style models
         ↓
GGUF conversion of bge-m3 loses/breaks the sparse & colbert heads
         ↓
llama.cpp tries to pool a partially broken model
         ↓
attention mask edge case → softmax(all -inf) → NaN
         ↓
Ollama can't JSON serialize NaN → 500 error
```

---

### The root cause in one line:
**bge-m3 is too complex for Ollama/llama.cpp to handle correctly** — it was never designed to run via llama.cpp, it was designed to run via HuggingFace's `FlagEmbedding` library.

---

### So your real options are:

1. **Use a simpler model in Ollama** that llama.cpp handles well — like `nomic-embed-text` or `mxbai-embed-large`

2. **Use bge-m3 via HuggingFace** — which requires solving the torch version issue

Which direction do you want to go?

In [4]:
import pandas as pd

eval_df = pd.read_csv("eval_dataset_v2.csv")

# Check for NaN values in any column
print("NaN counts per column:")
print(eval_df.isnull().sum())

# Find the problematic row
problem = eval_df[eval_df["question"].str.contains("primary application of Bhuva", na=False)]
print("\nProblematic row:")
print(problem[["question", "chunk_id", "ground_truth_text"]].to_string())

NaN counts per column:
question             0
chunk_id             0
source_file          0
page_number          0
ground_truth_text    0
dtype: int64

Problematic row:
                                                                                        question    chunk_id                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   

In [3]:
from langchain_huggingface import HuggingFaceEmbeddings

bge_embeddings = HuggingFaceEmbeddings(
    model_name="BAAI/bge-m3",
    model_kwargs={"device": "cuda"},  # uses your GPU
    encode_kwargs={"normalize_embeddings": True}
)

In [1]:
import torch
print(torch.__version__)  # 2.6.x+cu124

2.6.0+cu124


In [3]:
import shutil, os, time
from langchain_ollama import OllamaEmbeddings
from langchain_huggingface import HuggingFaceEmbeddings
from langchain_chroma import Chroma
from langchain_community.retrievers import BM25Retriever

# Load embedding models
nomic_embeddings = OllamaEmbeddings(model="nomic-embed-text")

bge_embeddings = HuggingFaceEmbeddings(
    model_name="BAAI/bge-m3",
    model_kwargs={"device": "cuda"},
    encode_kwargs={"normalize_embeddings": True}
)
print("Embedding models loaded")

# Delete both old vectorstores
for name in ["nomic", "bge"]:
    path = f"./chroma_db_{name}"
    if os.path.exists(path):
        shutil.rmtree(path)
        print(f"Deleted {path}")
time.sleep(2)

# Re-ingest both
embedding_map = {
    "nomic": nomic_embeddings,
    "bge"  : bge_embeddings,
}

vectorstores = {}
for name, embeddings in embedding_map.items():
    print(f"\nIngesting {name}...")
    start = time.time()
    vs = Chroma.from_documents(
        documents=clean_docs,
        embedding=embeddings,
        persist_directory=f"./chroma_db_{name}",
        collection_name=f"nrsc_docs_{name}"
    )
    elapsed = time.time() - start
    vectorstores[name] = vs
    print(f"{name} --> {vs._collection.count()} chunks in {int(elapsed//60)}m {int(elapsed%60)}s")

# Build retrievers
retrievers = {}
for name, vs in vectorstores.items():
    retrievers[f"{name}_similarity"] = vs.as_retriever(
        search_type="similarity",
        search_kwargs={"k": 10},
    )
    retrievers[f"{name}_mmr"] = vs.as_retriever(
        search_type="mmr",
        search_kwargs={"k": 10, "fetch_k": 50, "lambda_mult": 0.5},
    )
    retrievers[f"{name}_threshold"] = vs.as_retriever(
        search_type="similarity_score_threshold",
        search_kwargs={
            "score_threshold": 0.38 if name == "nomic" else 0.40,
            "k": 20
        },
    )

bm25_retriever = BM25Retriever.from_documents(clean_docs, bm25_variant="plus")
bm25_retriever.k = 10
retrievers["bm25"] = bm25_retriever

print(f"\nTotal retrievers: {len(retrievers)}")
for i, r in enumerate(retrievers):
    print(f"  {i}. {r}")

e:\The Leo Programmer\Internships\NRSC\Assignments\testing-models-huge-dataset\.venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Embedding models loaded

Ingesting nomic...
nomic --> 1487 chunks in 0m 22s

Ingesting bge...
bge --> 1487 chunks in 1m 3s

Total retrievers: 7
  0. nomic_similarity
  1. nomic_mmr
  2. nomic_threshold
  3. bge_similarity
  4. bge_mmr
  5. bge_threshold
  6. bm25


In [4]:
test_query = "What is the primary application of Bhuvan NextGen?"

for name, retriever in retrievers.items():
    try:
        results = retriever.invoke(test_query)
        top = results[0].page_content[:80].replace('\n',' ') if results else "NO RESULTS"
        print(f"[{name}] {len(results)} results --> {top}...")
    except Exception as e:
        print(f"[{name}] ERROR --> {e}")

[nomic_similarity] 10 results --> About NRSC  The National Remote Sensing Centre (NRSC) is one of the primary cent...
[nomic_mmr] 10 results --> About NRSC  The National Remote Sensing Centre (NRSC) is one of the primary cent...
[nomic_threshold] 20 results --> About NRSC  The National Remote Sensing Centre (NRSC) is one of the primary cent...
[bge_similarity] 10 results --> Bhuvan NextGen Inception  Natural resources information for 10 themes.  Bhuvan N...
[bge_mmr] 10 results --> Bhuvan NextGen Inception  Natural resources information for 10 themes.  Bhuvan N...
[bge_threshold] 8 results --> Bhuvan NextGen Inception  Natural resources information for 10 themes.  Bhuvan N...
[bm25] 10 results --> 2.4.3.2. Spectral Mixture Analysis  Many other means of quantifying vegetation c...


In [5]:
import os
if os.path.exists("eval_results_v2.csv"):
    os.remove("eval_results_v2.csv")
    print("Deleted partial results - starting fresh")
else:
    print("No partial results found")

No partial results found


In [6]:
import pandas as pd
import time
from tqdm import tqdm

eval_df = pd.read_csv("eval_dataset_v2.csv")
print(f"Total questions : {len(eval_df)}")
print(f"Total retrievers: {len(retrievers)}")
print(f"Total API calls : {len(eval_df) * len(retrievers)}")

try:
    results_df = pd.read_csv("eval_results_v2.csv")
    done_questions = set(zip(results_df["question"], results_df["retriever"]))
    print(f"Resuming -- {len(results_df)} results already saved")
except FileNotFoundError:
    results_df = pd.DataFrame()
    done_questions = set()
    print("Starting fresh")

print("\nStarting evaluation...")
print("=" * 50)

records = []
checkpoint_every = 100

for i, row in tqdm(eval_df.iterrows(), total=len(eval_df)):
    question     = row["question"]
    ground_truth = row["ground_truth_text"]
    chunk_id     = row["chunk_id"]

    for retriever_name, retriever in retrievers.items():

        if (question, retriever_name) in done_questions:
            continue

        try:
            docs = retriever.invoke(question)
            retrieved_texts = [doc.page_content for doc in docs]

            hit        = 0
            rank       = None
            n_returned = len(retrieved_texts)

            for r, text in enumerate(retrieved_texts):
                if text.strip() == ground_truth.strip():
                    hit  = 1
                    rank = r + 1
                    break

            records.append({
                "question"   : question,
                "chunk_id"   : chunk_id,
                "retriever"  : retriever_name,
                "hit"        : hit,
                "rank"       : rank,
                "n_returned" : n_returned,
            })

        except Exception as e:
            print(f"\n[ERROR] {retriever_name} | {question[:40]}: {e}")
            records.append({
                "question"   : question,
                "chunk_id"   : chunk_id,
                "retriever"  : retriever_name,
                "hit"        : 0,
                "rank"       : None,
                "n_returned" : 0,
            })
            continue

        time.sleep(0.05)

    if (i + 1) % checkpoint_every == 0:
        checkpoint_df = pd.DataFrame(records)
        combined = pd.concat([results_df, checkpoint_df], ignore_index=True)
        combined.to_csv("eval_results_v2.csv", index=False)
        results_df = combined
        records = []
        print(f"\n[{i+1}/{len(eval_df)}] Saved -- {len(results_df)} results")

if records:
    checkpoint_df = pd.DataFrame(records)
    combined = pd.concat([results_df, checkpoint_df], ignore_index=True)
    combined.to_csv("eval_results_v2.csv", index=False)
    print(f"\nDone! Total results saved: {len(combined)}")
else:
    print(f"\nDone! Total results saved: {len(results_df)}")

Total questions : 2878
Total retrievers: 7
Total API calls : 20146
Starting fresh

Starting evaluation...


  3%|▎         | 100/2878 [01:28<43:22,  1.07it/s]


[100/2878] Saved -- 700 results


  7%|▋         | 200/2878 [03:02<41:17,  1.08it/s]


[200/2878] Saved -- 1400 results


 10%|█         | 300/2878 [04:30<38:08,  1.13it/s]


[300/2878] Saved -- 2100 results


 14%|█▍        | 400/2878 [05:57<37:48,  1.09it/s]


[400/2878] Saved -- 2800 results


 17%|█▋        | 500/2878 [07:23<36:23,  1.09it/s]


[500/2878] Saved -- 3500 results


 21%|██        | 600/2878 [08:49<33:06,  1.15it/s]


[600/2878] Saved -- 4200 results


 24%|██▍       | 700/2878 [10:15<30:48,  1.18it/s]


[700/2878] Saved -- 4900 results


No relevant docs were retrieved using the relevance score threshold 0.4
 28%|██▊       | 800/2878 [11:41<31:23,  1.10it/s]


[800/2878] Saved -- 5600 results


 31%|███▏      | 900/2878 [13:07<28:17,  1.17it/s]


[900/2878] Saved -- 6300 results


 35%|███▍      | 1000/2878 [14:34<29:31,  1.06it/s]


[1000/2878] Saved -- 7000 results


 38%|███▊      | 1100/2878 [15:59<25:04,  1.18it/s]


[1100/2878] Saved -- 7700 results


 42%|████▏     | 1200/2878 [17:25<24:20,  1.15it/s]


[1200/2878] Saved -- 8400 results


 45%|████▌     | 1300/2878 [18:50<22:03,  1.19it/s]


[1300/2878] Saved -- 9100 results


 49%|████▊     | 1400/2878 [20:16<21:13,  1.16it/s]


[1400/2878] Saved -- 9800 results


No relevant docs were retrieved using the relevance score threshold 0.4
 52%|█████▏    | 1500/2878 [21:42<19:41,  1.17it/s]


[1500/2878] Saved -- 10500 results


 56%|█████▌    | 1600/2878 [23:07<19:18,  1.10it/s]


[1600/2878] Saved -- 11200 results


 59%|█████▉    | 1700/2878 [24:33<16:45,  1.17it/s]


[1700/2878] Saved -- 11900 results


 63%|██████▎   | 1800/2878 [26:05<23:14,  1.29s/it]


[1800/2878] Saved -- 12600 results


 66%|██████▌   | 1900/2878 [28:02<18:47,  1.15s/it]


[1900/2878] Saved -- 13300 results


 69%|██████▉   | 2000/2878 [29:54<21:57,  1.50s/it]


[2000/2878] Saved -- 14000 results


 73%|███████▎  | 2100/2878 [31:59<16:02,  1.24s/it]


[2100/2878] Saved -- 14700 results


 76%|███████▋  | 2200/2878 [33:57<14:23,  1.27s/it]


[2200/2878] Saved -- 15400 results


 80%|███████▉  | 2300/2878 [35:57<09:50,  1.02s/it]


[2300/2878] Saved -- 16100 results


 83%|████████▎ | 2400/2878 [37:35<09:11,  1.15s/it]


[2400/2878] Saved -- 16800 results


No relevant docs were retrieved using the relevance score threshold 0.4
 87%|████████▋ | 2500/2878 [39:10<06:20,  1.01s/it]


[2500/2878] Saved -- 17500 results


 90%|█████████ | 2600/2878 [40:49<04:28,  1.03it/s]


[2600/2878] Saved -- 18200 results


 94%|█████████▍| 2700/2878 [42:28<03:00,  1.01s/it]


[2700/2878] Saved -- 18900 results


 97%|█████████▋| 2793/2878 [44:05<01:41,  1.19s/it]No relevant docs were retrieved using the relevance score threshold 0.38
No relevant docs were retrieved using the relevance score threshold 0.4
 97%|█████████▋| 2800/2878 [44:14<01:41,  1.30s/it]


[2800/2878] Saved -- 19600 results


100%|██████████| 2878/2878 [45:50<00:00,  1.05it/s]



Done! Total results saved: 20146


In [7]:
# Metrics Computation

import pandas as pd

results_df = pd.read_csv("eval_results_v2.csv")

print(f"Total results: {len(results_df)}")
print(f"Retrievers   : {results_df['retriever'].unique().tolist()}")

metrics = {}

for retriever_name, group in results_df.groupby("retriever"):

    total    = len(group)
    hits     = group[group["hit"] == 1]

    # Hit Rate
    hit_rate = len(hits) / total

    # MRR — computed only on hits
    mrr = (1 / hits["rank"]).mean() if len(hits) > 0 else 0

    # Precision@k
    precision = group.apply(
        lambda r: 1 / r["n_returned"]
        if r["hit"] == 1 and r["n_returned"] > 0
        else 0, axis=1
    ).mean()

    # Coverage
    coverage = (group["n_returned"] > 0).sum() / total

    metrics[retriever_name] = {
        "Hit Rate"   : round(hit_rate, 4),
        "MRR"        : round(mrr, 4),
        "Precision@k": round(precision, 4),
        "Coverage"   : round(coverage, 4),
        "Total Qs"   : total,
    }

metrics_df = pd.DataFrame(metrics).T
metrics_df = metrics_df.sort_values("Hit Rate", ascending=False)
metrics_df.to_csv("eval_metrics_v2.csv")

print("\n=== FINAL RESULTS ===")
print(metrics_df.to_string())

Total results: 20146
Retrievers   : ['nomic_similarity', 'nomic_mmr', 'nomic_threshold', 'bge_similarity', 'bge_mmr', 'bge_threshold', 'bm25']

=== FINAL RESULTS ===
                  Hit Rate     MRR  Precision@k  Coverage  Total Qs
bge_similarity      0.9201  0.8349       0.0920     1.000    2878.0
bm25                0.8801  0.8816       0.0880     1.000    2878.0
nomic_threshold     0.8461  0.7199       0.0452     0.999    2878.0
bge_threshold       0.8200  0.8606       0.3199     0.910    2878.0
bge_mmr             0.8155  0.9052       0.0815     1.000    2878.0
nomic_similarity    0.8006  0.7568       0.0801     1.000    2878.0
nomic_mmr           0.6803  0.8449       0.0680     1.000    2878.0


### Fusion Retrieval

In [20]:
from langchain_classic.retrievers.ensemble import EnsembleRetriever

ensemble_retriever = EnsembleRetriever(
    retrievers=[
        retrievers["bge_similarity"],
        retrievers["bm25"]
    ],
    weights=[0.5, 0.5]
)

# Test
test_query = "What is Bhuvan and when did it start?"
results = ensemble_retriever.invoke(test_query)

print(f"Ensemble results: {len(results)}")
print(f"Top result: {results[0].page_content[:100]}")

Ensemble results: 13
Top result: About Bhuvan

Bhuvan, India’s national geospatial platform developed by ISRO and launched in 2009, i


In [19]:
def rrf_fusion(query, retriever1, retriever2, k=60, top_n=10):
    results1 = retriever1.invoke(query)
    results2 = retriever2.invoke(query)

    scores = {}
    doc_map = {}

    for rank, doc in enumerate(results1):
        key = doc.page_content.strip()
        scores[key] = scores.get(key, 0) + 1 / (k + rank + 1)
        doc_map[key] = doc

    for rank, doc in enumerate(results2):
        key = doc.page_content.strip()
        scores[key] = scores.get(key, 0) + 1 / (k + rank + 1)
        doc_map[key] = doc

    sorted_keys = sorted(scores.keys(), key=lambda x: scores[x], reverse=True)
    return [doc_map[k] for k in sorted_keys[:top_n]]

# Test
test_query = "What is Bhuvan and when did it start?"
results = rrf_fusion(test_query, retrievers["bge_similarity"], retrievers["bm25"])

print(f"Fusion results: {len(results)}")
print(f"Top result: {results[0].page_content[:100]}")

Fusion results: 10
Top result: About Bhuvan

Bhuvan, India’s national geospatial platform developed by ISRO and launched in 2009, i


In [21]:
# Fusion RRF and Fusion Ensemble

eval_df = pd.read_csv("eval_dataset_v2.csv")
print(f"Total questions: {len(eval_df)}")
print("Running fusion evaluation...")
print("=" * 50)

records = []

for i, row in tqdm(eval_df.iterrows(), total=len(eval_df)):
    question     = row["question"]
    ground_truth = row["ground_truth_text"]
    chunk_id     = row["chunk_id"]

    # Test both fusion approaches
    fusion_configs = {
        "fusion_rrf"      : lambda q: rrf_fusion(q, retrievers["bge_similarity"], retrievers["bm25"]),
        "fusion_ensemble" : lambda q: ensemble_retriever.invoke(q),
    }

    for retriever_name, retriever_fn in fusion_configs.items():
        try:
            docs           = retriever_fn(question)
            retrieved_texts = [doc.page_content for doc in docs]

            hit        = 0
            rank       = None
            n_returned = len(retrieved_texts)

            for r, text in enumerate(retrieved_texts):
                if text.strip() == ground_truth.strip():
                    hit  = 1
                    rank = r + 1
                    break

            records.append({
                "question"   : question,
                "chunk_id"   : chunk_id,
                "retriever"  : retriever_name,
                "hit"        : hit,
                "rank"       : rank,
                "n_returned" : n_returned,
            })

        except Exception as e:
            print(f"\n[ERROR] {retriever_name} | {question[:40]}: {e}")
            records.append({
                "question"   : question,
                "chunk_id"   : chunk_id,
                "retriever"  : retriever_name,
                "hit"        : 0,
                "rank"       : None,
                "n_returned" : 0,
            })
            continue

    if (i + 1) % 100 == 0:
        print(f"[{i+1}/{len(eval_df)}] processed...")

# Save
fusion_df = pd.DataFrame(records)
fusion_df.to_csv("eval_results_fusion_v2.csv", index=False)
print(f"\nDone! Total results: {len(fusion_df)}")

# Compute metrics
print("\n=== FUSION RESULTS ===")
print(f"{'Retriever':<20} {'Hit Rate':>10} {'MRR':>10} {'Precision':>10} {'Coverage':>10}")

for retriever_name, group in fusion_df.groupby("retriever"):
    hits      = group[group["hit"] == 1]
    hit_rate  = len(hits) / len(group)
    mrr       = (1 / hits["rank"]).mean() if len(hits) > 0 else 0
    precision = group.apply(
        lambda r: 1 / r["n_returned"] if r["hit"] == 1 and r["n_returned"] > 0 else 0,
        axis=1
    ).mean()
    coverage  = (group["n_returned"] > 0).sum() / len(group)
    print(f"{retriever_name:<20} {hit_rate:>10.4f} {mrr:>10.4f} {precision:>10.4f} {coverage:>10.4f}")

print("\n=== COMPARISON WITH INDIVIDUAL RETRIEVERS ===")
print(f"{'Retriever':<20} {'Hit Rate':>10} {'MRR':>10}")
print(f"{'bge_similarity':<20} {0.9201:>10.4f} {0.8349:>10.4f}")
print(f"{'bm25':<20} {0.8801:>10.4f} {0.8816:>10.4f}")

Total questions: 2878
Running fusion evaluation...


  4%|▎         | 103/2878 [00:14<02:49, 16.33it/s]

[100/2878] processed...


  7%|▋         | 201/2878 [00:30<08:27,  5.27it/s]

[200/2878] processed...


 10%|█         | 300/2878 [00:50<08:39,  4.96it/s]

[300/2878] processed...


 14%|█▍        | 400/2878 [01:11<08:46,  4.71it/s]

[400/2878] processed...


 17%|█▋        | 503/2878 [01:31<03:39, 10.84it/s]

[500/2878] processed...


 21%|██        | 601/2878 [01:38<02:26, 15.51it/s]

[600/2878] processed...


 24%|██▍       | 703/2878 [01:44<02:13, 16.30it/s]

[700/2878] processed...


 28%|██▊       | 803/2878 [01:50<02:02, 16.87it/s]

[800/2878] processed...


 31%|███▏      | 903/2878 [01:57<02:07, 15.54it/s]

[900/2878] processed...


 35%|███▍      | 1003/2878 [02:03<01:49, 17.10it/s]

[1000/2878] processed...


 38%|███▊      | 1103/2878 [02:09<01:48, 16.29it/s]

[1100/2878] processed...


 42%|████▏     | 1203/2878 [02:15<01:36, 17.44it/s]

[1200/2878] processed...


 45%|████▌     | 1303/2878 [02:21<01:41, 15.52it/s]

[1300/2878] processed...


 49%|████▊     | 1403/2878 [02:28<01:27, 16.91it/s]

[1400/2878] processed...


 52%|█████▏    | 1503/2878 [02:34<01:23, 16.41it/s]

[1500/2878] processed...


 56%|█████▌    | 1601/2878 [02:40<01:12, 17.73it/s]

[1600/2878] processed...


 59%|█████▉    | 1703/2878 [02:46<01:13, 16.06it/s]

[1700/2878] processed...


 63%|██████▎   | 1803/2878 [02:52<01:10, 15.22it/s]

[1800/2878] processed...


 66%|██████▌   | 1903/2878 [02:58<00:56, 17.13it/s]

[1900/2878] processed...


 70%|██████▉   | 2003/2878 [03:04<00:52, 16.62it/s]

[2000/2878] processed...


 73%|███████▎  | 2103/2878 [03:10<00:47, 16.46it/s]

[2100/2878] processed...


 77%|███████▋  | 2203/2878 [03:16<00:41, 16.38it/s]

[2200/2878] processed...


 80%|████████  | 2303/2878 [03:22<00:35, 16.19it/s]

[2300/2878] processed...


 83%|████████▎ | 2403/2878 [03:28<00:27, 17.09it/s]

[2400/2878] processed...


 87%|████████▋ | 2503/2878 [03:34<00:22, 16.94it/s]

[2500/2878] processed...


 90%|█████████ | 2603/2878 [03:40<00:15, 18.15it/s]

[2600/2878] processed...


 94%|█████████▍| 2703/2878 [03:46<00:10, 17.09it/s]

[2700/2878] processed...


 97%|█████████▋| 2803/2878 [03:52<00:04, 17.55it/s]

[2800/2878] processed...


100%|██████████| 2878/2878 [03:56<00:00, 12.15it/s]


Done! Total results: 5756

=== FUSION RESULTS ===
Retriever              Hit Rate        MRR  Precision   Coverage
fusion_ensemble          0.9649     0.8395     0.0577     1.0000
fusion_rrf               0.9559     0.8468     0.0956     1.0000

=== COMPARISON WITH INDIVIDUAL RETRIEVERS ===
Retriever              Hit Rate        MRR
bge_similarity           0.9201     0.8349
bm25                     0.8801     0.8816
